### PR TITLE
Thread Decorator

### PR DESCRIPTION
**Pull request by Velin92**
Added a thread decorator and fixed the overall behaviour of images, videos and locations when in replies
Screenshot 2023-09-13 at 17 47 10
Screenshot 2023-09-13 at 17 47 04
Screenshot 2023-09-13 at 17 46 58
Simulator Screenshot - iPhone 14 - 2023-09-13 at 17 46 22
Simulat
![Simulator Screenshot - iPhone 14 - 2023-09-13 at 17 48 35
or Screenshot - iPhone 14 - 2023-09-13 at 17 46 17](https://github.com/vector-im/element-x-ios/assets/34335419/de303edc-2d8e-4697-83e9-27ca1fa9faf2)

Simulator Screenshot - iPhone 14 - 2023-09-13 at 17 48 39
Simulator Screenshot - iPhone 14 - 2023-09-13 at 17 48 39

Screenshot 2023-09-13 at 15 52 19

### CODE REVIEW

**Comment by callumu** [comment]
This all looks good to me 👍

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/LocationRoomTimelineView.swift [hunk 1]
```diff
@@ -99,14 +107,21 @@ struct LocationRoomTimelineView_Previews: PreviewProvider {
         LocationRoomTimelineView(timelineItem: .init(id: .random,
                                                      timestamp: "Now",
                                                      isOutgoing: false,
-                                                     isEditable: false,
+                                                     isEditable: false, isThreaded: false,
```
Comment (alfogrillo) [inline ElementX/Sources/Screens/RoomScreen/View/Timeline/LocationRoomTimelineView.swift:110]:
```suggestion
                                                     isEditable: false, 
                                                     isThreaded: false,
```

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/ReadMarkerRoomTimelineView.swift [hunk 2]
```diff
@@ -46,7 +46,7 @@ struct ReadMarkerRoomTimelineView_Previews: PreviewProvider {
             RoomTimelineItemView(viewState: .init(type: .text(.init(id: .init(timelineID: ""),
                                                                     timestamp: "",
                                                                     isOutgoing: true,
-                                                                    isEditable: false,
+                                                                    isEditable: false, isThreaded: false,
```
Comment (alfogrillo) [inline ElementX/Sources/Screens/RoomScreen/View/Timeline/ReadMarkerRoomTimelineView.swift:49]:
```suggestion
                                                                    isEditable: false, 
                                                                    isThreaded: false,
```

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/LocationRoomTimelineView.swift [hunk 3]
```diff
@@ -20,6 +20,13 @@ struct LocationRoomTimelineView: View {
     let timelineItem: LocationRoomTimelineItem
     @Environment(\.timelineStyle) var timelineStyle
     
+    private var mapCornerRadius: CGFloat {
+        if timelineStyle.isBubbles, timelineItem.replyDetails != nil || timelineItem.isThreaded {
+            return 8
+        }
+        return 0
+    }
```
Comment (alfogrillo) [inline ElementX/Sources/Screens/RoomScreen/View/Timeline/LocationRoomTimelineView.swift:28]: Can you explain why we need to handle the corner radius internally for location events?
They should work pretty much like normal images (except when a location has a description in which case the UI is different)

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/ImageRoomTimelineView.swift [hunk 4]
```diff
@@ -28,8 +28,8 @@ struct ImageRoomTimelineView: View {
                           imageProvider: context.imageProvider) {
                 placeholder
             }
-            .frame(maxHeight: min(300, max(100, timelineItem.content.height ?? .infinity)))
             .aspectRatio(timelineItem.content.aspectRatio, contentMode: .fit)
+            .frame(maxHeight: min(300, max(100, timelineItem.content.height ?? .infinity)))
```
Comment (alfogrillo) [inline ElementX/Sources/Screens/RoomScreen/View/Timeline/ImageRoomTimelineView.swift:32]: What this fixed?

File: ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift [hunk 5]
```diff
@@ -516,8 +516,8 @@ class RoomScreenViewModel: RoomScreenViewModelType, RoomScreenViewModelProtocol
         
         var actions: [TimelineItemMenuAction] = []
 
-        if item.isMessage {
-            actions.append(.reply)
+        if let messageitem = item as? EventBasedMessageTimelineItemProtocol {
+            actions.append(.reply(isThread: messageitem.isThreaded))
```
[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift:520]: Not Mauro, but I'm going to get this over the finish line for now: `isThreaded` will only be true for items inside the thread, not the thread root. I assume this is why.

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/ImageRoomTimelineView.swift (hunk 4 above)
Comment (pixlwave) [reply ElementX/Sources/Screens/RoomScreen/View/Timeline/ImageRoomTimelineView.swift:32]: There was an issue where when there was a reply at the top, the message's content didn't always fill the bubble properly:

simulator_screenshot_AF5CBFC8-F83C-479D-955F-8A03DDFBA3D5

File: ElementX/Sources/Other/SwiftUI/Layout/CustomLayoutLabelStyle.swift [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import SwiftUI
+
+struct CustomLayoutLabelStyle: LabelStyle {
+    let spacing: CGFloat
```
[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Other/SwiftUI/Layout/CustomLayoutLabelStyle.swift:20]: Given that basically the point of this label style is to customise the spacing over Apple's default value, I feel like this is enough to not provide a default value and make us be explicit about the spacing required.

File: ElementX/Sources/Other/SwiftUI/Layout/CustomLayoutLabelStyle.swift [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    var alignment = VerticalAlignment.center
+    var isInverted = false
```
[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Other/SwiftUI/Layout/CustomLayoutLabelStyle.swift:22]: Agreed, adding an enum.

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/LocationRoomTimelineView.swift (hunk 3 above)
Comment (pixlwave) [reply ElementX/Sources/Screens/RoomScreen/View/Timeline/LocationRoomTimelineView.swift:28]: Honestly I'm not 100% sure on this one, the bubble styler has got a lot more gnarly recently it seems.

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/ImageRoomTimelineView.swift (hunk 4 above)
Comment (pixlwave) [reply ElementX/Sources/Screens/RoomScreen/View/Timeline/ImageRoomTimelineView.swift]: I've extracted this out into a `timelineMediaFrame` modifier so we have the implementation shared everywhere too.

File: ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (alfogrillo) [reply ElementX/Sources/Screens/RoomScreen/RoomScreenViewModel.swift:521]: > will only be true for items inside the thread, not the thread root.

Ok, we should be good then! 👍🏻

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/LocationRoomTimelineView.swift (hunk 3 above)
Comment (alfogrillo) [reply ElementX/Sources/Screens/RoomScreen/View/Timeline/LocationRoomTimelineView.swift:28]: I moved this logic with the other cases in `TimelineItemBubbledStylerView.swift`

like this:
```swift
    var contentCornerRadius: CGFloat {
        if let message = self as? EventBasedMessageTimelineItemProtocol,
           message is ImageRoomTimelineItem || message is VideoRoomTimelineItem || message is LocationRoomTimelineItem {
            return message.replyDetails != nil || message.isThreaded ? 8 : .zero
        }
        return .zero
    }
```
And the result looks the same to me.
Can you please verify if it's good?

Simulator Screenshot - iPhone 14 - 2023-09-15 at 09 00 17

File: ElementX/Sources/Other/SwiftUI/Layout/CustomLayoutLabelStyle.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (alfogrillo) [reply ElementX/Sources/Other/SwiftUI/Layout/CustomLayoutLabelStyle.swift:27]: I thought the main point was to be able to invert the image and the text.
But yes, it makes sense 👍🏻

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/LocationRoomTimelineView.swift (hunk 3 above)
Comment (alfogrillo) [reply ElementX/Sources/Screens/RoomScreen/View/Timeline/LocationRoomTimelineView.swift:28]: (btw I think that writing   `var contentCornerRadius: CGFloat` as a `switch` would be much easier to be read and also easier to be extended in the future)

**Review by alfogrillo — APPROVED** [review]
LGTM

File: ElementX/Sources/Screens/RoomScreen/View/Timeline/LocationRoomTimelineView.swift (hunk 3 above)
Comment (alfogrillo) [reply ElementX/Sources/Screens/RoomScreen/View/Timeline/LocationRoomTimelineView.swift]: Discussed offline with Doug

**Comment by callumu** [comment]
@Velin92 small thing for when you have the time: can you change the 'Reply in thread' text when swiping on a message to have 'Reply in' on the first line and 'thread' on the second line?

**Pull request merged**